include/zephyr/drivers/retained_mem.h

```diff
@@ -0,0 +1,194 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Public API for retained memory drivers
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_RETAINED_MEM_
+#define ZEPHYR_INCLUDE_DRIVERS_RETAINED_MEM_
+
+#include <stdint.h>
+#include <stddef.h>
+#include <sys/types.h>
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/types.h>
+#include <zephyr/sys/math_extras.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Retained memory driver interface
+ * @defgroup retained_mem_interface Retamined memory driver interface
+ * @ingroup io_interfaces
+ * @{
+ */
+
+/**
+ * @typedef	retained_mem_size_api
+ * @brief	Callback API to get size of retained memory area.
+ * See retained_mem_size() for argument description.
+ */
+typedef ssize_t (*retained_mem_size_api)(const struct device *dev);
```

Comment:
Just for errors, both drivers here cannot return an error because they're direct memory or registers but there might be a driver in the future that has a communication method that is not guaranteed to work e.g. IPC or I2C, so left the possibility of returning an error if it cannot be checked